DataFormats/ParticleFlowReco/interface/PFRecHit.h

```diff
@@ -74,81 +71,71 @@ namespace reco {
 
-    void setEnergy( double energy) { energy_ = energy; }
+    void setEnergy( float energy) { energy_ = energy; }
 
-    void calculatePositionREP();
 
-    void addNeighbour(short x,short y, short z,const PFRecHitRef&);
-    const PFRecHitRef getNeighbour(short x,short y, short z);
+    void addNeighbour(short x,short y, short z, unsigned int);
+    unsigned int getNeighbour(short x,short y, short z) const;
     void setTime( double time) { time_ = time; }
     void setDepth( int depth) { depth_ = depth; }
     void clearNeighbours() {
       neighbours_.clear();
+      neighbourInfos_.clear();
+      neighbours4_ = neighbours8_ = 0;
     }
 
-    const PFRecHitRefVector& neighbours4() const {
-      return neighbours4_;
+    Neighbours neighbours4() const {
+      return buildNeighbours(neighbours4_);
     }
-    const PFRecHitRefVector& neighbours8() const {
-      return neighbours8_;
+    Neighbours neighbours8() const {
+	return buildNeighbours(neighbours8_);
     }
 
-    const PFRecHitRefVector& neighbours() const {
-      return neighbours_;
+    Neighbours neighbours() const {
+      return buildNeighbours(neighbours_.size());
     }
 
     const std::vector<unsigned short>& neighbourInfos() {
       return neighbourInfos_;
     }
 
 
-    void      setNWCorner( double posx, double posy, double posz );
-    void      setSWCorner( double posx, double posy, double posz );
-    void      setSECorner( double posx, double posy, double posz );
-    void      setNECorner( double posx, double posy, double posz );
-
+    /// calo cell
+    CaloCellGeometry const & caloCell() const { return  *caloCell_; }
+    bool hasCaloCell() const { return caloCell_; }
+    
     /// rechit detId
     unsigned detId() const {return detId_;}
 
     /// rechit layer
     PFLayer::Layer layer() const { return layer_; }
 
     /// rechit energy
-    double energy() const { return energy_; }
+    float energy() const { return energy_; }
 
 
     /// timing for cleaned hits
-    double time() const { return time_; }
+    float time() const { return time_; }
 
     /// depth for segemntation
     int  depth() const { return depth_; }
 
     /// rechit momentum transverse to the beam, squared.
     double pt2() const { return energy_ * energy_ *
-			   ( position_.X()*position_.X() + 
-			     position_.Y()*position_.Y() ) / 
-			   ( position_.X()*position_.X() +
-			     position_.Y()*position_.Y() + 
-			     position_.Z()*position_.Z()) ; }
+	( position().perp2()/ position().mag2());}
 
 
     /// rechit cell centre x, y, z
-    const math::XYZPoint& position() const { return position_; }
-
-    const REPPoint& positionREP() const { return positionrep_; }
-
-
-    /// rechit cell axis x, y, z
-    const math::XYZVector& getAxisXYZ() const { return axisxyz_; }    
+    PositionType const & position() const { return caloCell().getPosition().basicVector(); }
+    
+    RhoEtaPhi const &  positionREP() const { return caloCell().repPos(); }
```

Comment:
if needed
here is not hidden
http://cmslxr.fnal.gov/dxr/CMSSW_8_0_5/source/DataFormats/TrackingRecHit/interface/TrackingRecHit.h#119